SSWC/doc/man/man3/SpaceObject.3
.TH "SpaceObject" 3 "Mon May 9 2016" "Version 0.1" "MissionsVisualizer" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SpaceObject \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SpaceObject\&.h>\fP
.PP
Inherited by \fBSpaceBody\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSpaceObject\fP (long \fBspiceId\fP, const \fBstd::string\fP &\fBname\fP='')"
.br
.ti -1c
.RI "\fBSpaceObject\fP (const \fBstd::string\fP &spiceName, const \fBstd::string\fP &\fBname\fP='')"
.br
.ti -1c
.RI "virtual \fB~SpaceObject\fP ()"
.br
.ti -1c
.RI "virtual \fBSpaceObject\fP * \fBClone\fP () const "
.br
.ti -1c
.RI "long \fBGetSpiceId\fP () const "
.br
.ti -1c
.RI "\fBstd::string\fP \fBGetSpiceName\fP () const "
.br
.ti -1c
.RI "const \fBstd::string\fP & \fBGetName\fP () const "
.br
.ti -1c
.RI "\fBVector3T\fP< \fBLength\fP > \fBGetPosition\fP (const \fBDate\fP &\fBt\fP, const \fBSpaceObject\fP &relativeTo, const \fBFrame\fP &frame) const "
.br
.ti -1c
.RI "\fBVector3T\fP< \fBLength\fP > \fBGetPosition\fP (const \fBDate\fP &\fBt\fP, const \fBFrame\fP &frame) const "
.br
.ti -1c
.RI "\fBVector3T\fP< \fBVelocity\fP > \fBGetVelocity\fP (const \fBDate\fP &\fBt\fP, const \fBSpaceObject\fP &relativeTo, const \fBFrame\fP &frame) const "
.br
.ti -1c
.RI "\fBVector3T\fP< \fBVelocity\fP > \fBGetVelocity\fP (const \fBDate\fP &\fBt\fP, const \fBFrame\fP &frame) const "
.br
.ti -1c
.RI "\fBWindow\fP \fBGetCoverage\fP () const "
.br
.ti -1c
.RI "bool \fBIsBarycenter\fP () const "
.br
.ti -1c
.RI "bool \fBIsPlanetaryBarycenter\fP () const "
.br
.ti -1c
.RI "bool \fBIsSSB\fP () const "
.br
.ti -1c
.RI "bool \fBIsPlanet\fP () const "
.br
.ti -1c
.RI "bool \fBIsMoon\fP () const "
.br
.ti -1c
.RI "bool \fBIsBody\fP () const "
.br
.ti -1c
.RI "bool \fBIsSun\fP () const "
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBValidateId\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsBarycenter\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsPlanetaryBarycenter\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsSSB\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsPlanet\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsMoon\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsBody\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static bool \fBIsSun\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static std::vector< long > \fBFindChildObjectIds\fP (long \fBid\fP, bool includeMassCenter=true)"
.br
.ti -1c
.RI "static long \fBFindChildMassCenterId\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static std::vector< long > \fBFindMoonIds\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static long \fBFindParentObjectId\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static std::vector< long > \fBGetLoadedSpkIds\fP ()"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const \fBSpaceObject\fP \fBSSB\fP = \fBSpaceObject\fP(\fBSSB_SPICE_ID\fP, 'Solar System Barycenter')"
.br
.ti -1c
.RI "static const \fBSpaceObject\fP \fBSun\fP = \fBSpaceObject\fP(\fBSUN_SPICE_ID\fP, 'Sun')"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "long \fBspiceId\fP"
.br
.ti -1c
.RI "\fBstd::string\fP \fBname\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 20 of file SpaceObject\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SpaceObject::SpaceObject (longspiceId, const \fBstd::string\fP &name = \fC''\fP)"

.PP
Definition at line 3 of file SpaceObject\&.cpp\&.
.SS "SpaceObject::SpaceObject (const \fBstd::string\fP &spiceName, const \fBstd::string\fP &name = \fC''\fP)"

.PP
Definition at line 8 of file SpaceObject\&.cpp\&.
.SS "SpaceObject::~SpaceObject ()\fC [virtual]\fP"

.PP
Definition at line 37 of file SpaceObject\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBSpaceObject\fP * SpaceObject::Clone () const\fC [virtual]\fP"

.PP
Reimplemented in \fBSpaceBody\fP\&.
.PP
Definition at line 42 of file SpaceObject\&.cpp\&.
.SS "long SpaceObject::FindChildMassCenterId (longid)\fC [static]\fP"

.PP
Definition at line 265 of file SpaceObject\&.cpp\&.
.SS "std::vector< long > SpaceObject::FindChildObjectIds (longid, boolincludeMassCenter = \fCtrue\fP)\fC [static]\fP"

.PP
Definition at line 234 of file SpaceObject\&.cpp\&.
.SS "std::vector< long > SpaceObject::FindMoonIds (longid)\fC [static]\fP"

.PP
Definition at line 275 of file SpaceObject\&.cpp\&.
.SS "long SpaceObject::FindParentObjectId (longid)\fC [static]\fP"

.PP
Definition at line 285 of file SpaceObject\&.cpp\&.
.SS "\fBWindow\fP SpaceObject::GetCoverage () const"

.PP
Definition at line 118 of file SpaceObject\&.cpp\&.
.SS "std::vector< long > SpaceObject::GetLoadedSpkIds ()\fC [static]\fP"

.PP
Definition at line 301 of file SpaceObject\&.cpp\&.
.SS "const \fBstd::string\fP & SpaceObject::GetName () const"

.PP
Definition at line 61 of file SpaceObject\&.cpp\&.
.SS "\fBVector3T\fP< \fBLength\fP > SpaceObject::GetPosition (const \fBDate\fP &t, const \fBSpaceObject\fP &relativeTo, const \fBFrame\fP &frame) const"

.PP
Definition at line 66 of file SpaceObject\&.cpp\&.
.SS "\fBVector3T\fP< \fBLength\fP > SpaceObject::GetPosition (const \fBDate\fP &t, const \fBFrame\fP &frame) const"

.PP
Definition at line 84 of file SpaceObject\&.cpp\&.
.SS "long SpaceObject::GetSpiceId () const"

.PP
Definition at line 47 of file SpaceObject\&.cpp\&.
.SS "\fBstd::string\fP SpaceObject::GetSpiceName () const"

.PP
Definition at line 52 of file SpaceObject\&.cpp\&.
.SS "\fBVector3T\fP< \fBVelocity\fP > SpaceObject::GetVelocity (const \fBDate\fP &t, const \fBSpaceObject\fP &relativeTo, const \fBFrame\fP &frame) const"

.PP
Definition at line 91 of file SpaceObject\&.cpp\&.
.SS "\fBVector3T\fP< \fBVelocity\fP > SpaceObject::GetVelocity (const \fBDate\fP &t, const \fBFrame\fP &frame) const"

.PP
Definition at line 111 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsBarycenter () const"

.PP
Definition at line 134 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsBarycenter (longid)\fC [static]\fP"

.PP
Definition at line 178 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsBody () const"

.PP
Definition at line 159 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsBody (longid)\fC [static]\fP"

.PP
Definition at line 218 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsMoon () const"

.PP
Definition at line 154 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsMoon (longid)\fC [static]\fP"

.PP
Definition at line 210 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsPlanet () const"

.PP
Definition at line 149 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsPlanet (longid)\fC [static]\fP"

.PP
Definition at line 202 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsPlanetaryBarycenter () const"

.PP
Definition at line 139 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsPlanetaryBarycenter (longid)\fC [static]\fP"

.PP
Definition at line 186 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsSSB () const"

.PP
Definition at line 144 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsSSB (longid)\fC [static]\fP"

.PP
Definition at line 194 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsSun () const"

.PP
Definition at line 164 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::IsSun (longid)\fC [static]\fP"

.PP
Definition at line 226 of file SpaceObject\&.cpp\&.
.SS "bool SpaceObject::ValidateId (longid)\fC [static]\fP"

.PP
Definition at line 169 of file SpaceObject\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBstd::string\fP SpaceObject::name\fC [protected]\fP"

.PP
Definition at line 72 of file SpaceObject\&.h\&.
.SS "long SpaceObject::spiceId\fC [protected]\fP"

.PP
Definition at line 71 of file SpaceObject\&.h\&.
.SS "const \fBSpaceObject\fP SpaceObject::SSB = \fBSpaceObject\fP(\fBSSB_SPICE_ID\fP, 'Solar System Barycenter')\fC [static]\fP"

.PP
Definition at line 75 of file SpaceObject\&.h\&.
.SS "const \fBSpaceObject\fP SpaceObject::Sun = \fBSpaceObject\fP(\fBSUN_SPICE_ID\fP, 'Sun')\fC [static]\fP"

.PP
Definition at line 76 of file SpaceObject\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for MissionsVisualizer from the source code\&.


SSWC/src/CSpice/SpaceObject.cpp
#include "SpaceObject.h"

SpaceObject::SpaceObject(long spiceId, const std::string& name)
{
	Construct(spiceId, name);
}

SpaceObject::SpaceObject(const std::string& spiceName, const std::string& name)
{
	SpiceInt spiceId;
	SpiceBoolean found;
	CSPICE_ASSERT( bodn2c_c(spiceName.c_str(), &spiceId, &found) );

	if(!found)
		CSpiceUtil::SignalError("Cannot translate " + spiceName + " into ID code");

	Construct(spiceId, name);
}

void SpaceObject::Construct(long spiceId, const std::string& name)
{
	if(!ValidateId(spiceId))
		CSpiceUtil::SignalError("No such CSpice object ID code is defined: " + std::to_string(spiceId));

	this->spiceId = spiceId;

	if(name != "")
	{
		this->name = name;
	}
	else
	{
		this->name = GetSpiceName();
	}
}

SpaceObject::~SpaceObject()
{

}

SpaceObject* SpaceObject::Clone() const
{
	return new SpaceObject(*this);
}

long SpaceObject::GetSpiceId() const
{
	return spiceId;
}

std::string SpaceObject::GetSpiceName() const
{
	char objName[OBJECT_NAME_MAX_LENGTH];
	SpiceBoolean found;
	CSPICE_ASSERT(bodc2n_c(spiceId, OBJECT_NAME_MAX_LENGTH, objName, &found));

	return std::string(objName);
}

const std::string& SpaceObject::GetName() const
{
	return name;
}

Vector3T<Length> SpaceObject::GetPosition(const Date& t, const SpaceObject& relativeTo, const Frame& frame) const
{
	double etTime = t.AsDouble();
	long observerId = relativeTo.GetSpiceId();
	std::string frameName = frame.GetSpiceName();

	double position[3];
	double lt;

	CSPICE_ASSERT(spkgps_c(this->spiceId, etTime, frameName.c_str(), observerId, position, &lt));

	Length pos[3];
	for(int i = 0; i < 3; i++)
		pos[i] = Length(position[i], Units::Metric::kilometers);

	return Vector3T<Length>(pos);
}

Vector3T<Length> SpaceObject::GetPosition(const Date& t, const Frame& frame) const
{
	Frame::FrameInfo frInfo = frame.GetFrameInfo();
	
	return GetPosition(t, frame.GetCenterObject(), frame);
}

Vector3T<Velocity> SpaceObject::GetVelocity(const Date& t, const SpaceObject& relativeTo, const Frame& frame) const
{
	double etTime = t.AsDouble();
	long observerId = relativeTo.GetSpiceId();
	std::string frameName = frame.GetSpiceName();

	double state[6];
	double lt;

	CSPICE_ASSERT(spkgeo_c(this->spiceId, etTime, frameName.c_str(), observerId, state, &lt));

	Velocity vel[3];
	for(int i = 0; i < 3; i++)
	{
		vel[i] = Velocity(state[i + 3], Units::Metric::kmps);
	}

	return Vector3T<Velocity>(vel);
}

Vector3T<Velocity> SpaceObject::GetVelocity(const Date& t, const Frame& frame) const
{
	Frame::FrameInfo frInfo = frame.GetFrameInfo();

	return GetVelocity(t, SpaceObject(frInfo.centerId), frame);
}

Window SpaceObject::GetCoverage() const
{
	std::vector<KernelData> kernels = CSpiceUtil::GetLoadedKernels("SPK");

	Window coverage;

	for(size_t i = 0; i < kernels.size(); i++)
	{
		std::string file = kernels[i].filename;

		CSPICE_ASSERT( spkcov_c(file.c_str(), this->spiceId, &coverage.GetSpiceCell()) );
	}

	return coverage;
}

bool SpaceObject::IsBarycenter() const
{
	return SpaceObject::IsBarycenter(spiceId);
}

bool SpaceObject::IsPlanetaryBarycenter() const
{
	return SpaceObject::IsPlanetaryBarycenter(spiceId);
}

bool SpaceObject::IsSSB() const
{
	return SpaceObject::IsSSB(spiceId);
}

bool SpaceObject::IsPlanet() const
{
	return SpaceObject::IsPlanet(spiceId);
}

bool SpaceObject::IsMoon() const
{
	return SpaceObject::IsMoon(spiceId);
}

bool SpaceObject::IsBody() const
{
	return SpaceObject::IsBody(spiceId);
}

bool SpaceObject::IsSun() const
{
	return SpaceObject::IsSun(spiceId);
}

bool SpaceObject::ValidateId(long id)
{
	char objName[OBJECT_NAME_MAX_LENGTH];
	SpiceBoolean found;
	CSPICE_ASSERT(bodc2n_c(id, OBJECT_NAME_MAX_LENGTH, objName, &found));

	return found != SPICEFALSE;
}

bool SpaceObject::IsBarycenter(long id)
{
	if(ValidateId(id) == false)
		return false;

	return IsPlanetaryBarycenter(id) || IsSSB(id);
}

bool SpaceObject::IsPlanetaryBarycenter(long id)
{
	if(ValidateId(id) == false)
		return false;

	return (id > 0 && id < 10);
}

bool SpaceObject::IsSSB(long id)
{
	if(ValidateId(id) == false)
		return false;

	return id == SSB_SPICE_ID;
}

bool SpaceObject::IsPlanet(long id)
{
	if(ValidateId(id) == false)
		return false;

	return (id > 100 && id < 1000 && (id % 100) == 99);
}

bool SpaceObject::IsMoon(long id)
{
	if(ValidateId(id) == false)
		return false;

	return IsBody(id) && !IsPlanet(id) && !IsSun(id);
}

bool SpaceObject::IsBody(long id)
{
	if(ValidateId(id) == false)
		return false;

	return ((id > 100 && id < 1000) || IsSun(id));
}

bool SpaceObject::IsSun(long id)
{
	if(ValidateId(id) == false)
		return false;

	return id == SUN_SPICE_ID;
}

std::vector<long> SpaceObject::FindChildObjectIds(long id, bool includeMain)
{
	std::vector<long> ids;

	if(id == SSB_SPICE_ID)
	{
		for(int i = 1; i <= 10; i++)
		{
			if(ValidateId(i))
				ids.push_back(i);
		}
	}
	else if(IsPlanetaryBarycenter(id))
	{
		int upperSatelliteId;
		if(includeMain)
			upperSatelliteId = 99; // X99 - id of a most massive body in the X-th barycenter
		else
			upperSatelliteId = 98;

		for(int i = 1; i <= upperSatelliteId; i++)
		{
			long satelliteId = 100 * id + i;
			if(ValidateId(satelliteId))
				ids.push_back(satelliteId);
		}
	}

	return ids;
}

long SpaceObject::FindChildMassCenterId(long id)
{
	if(IsPlanetaryBarycenter(id))
		return id * 100 + 99;
	else if(IsSSB(id))
		return SUN_SPICE_ID;
	else
		return id;
}

std::vector<long> SpaceObject::FindMoonIds(long id)
{
	long barycenterId = SpaceObject::FindParentObjectId(id);

	if(!SpaceObject::IsPlanetaryBarycenter(barycenterId))
		CSpiceUtil::SignalError("FindMoonIds failed: 'id' argument is not a valid planet id");

	return FindChildObjectIds(barycenterId, false);
}

long SpaceObject::FindParentObjectId(long id)
{
	if(id == SUN_SPICE_ID || IsPlanetaryBarycenter(id))
	{
		return SSB.GetSpiceId();
	}
	else if(IsBody(id))
	{
		return id / 100;
	}
	else
	{
		return id;
	}
}

std::vector<long> SpaceObject::GetLoadedSpkIds()
{
	SPICEINT_CELL(cell, CELL_SIZE_LARGE);
	std::vector<KernelData> kernels = CSpiceUtil::GetLoadedKernels("SPK");
	for(size_t i = 0; i < kernels.size(); i++)
	{
		CSPICE_ASSERT(spkobj_c(kernels[i].filename.c_str(), &cell));
	}

	return CSpiceUtil::IntCellToVector(cell);
}

const SpaceObject SpaceObject::SSB = SpaceObject(SSB_SPICE_ID, "Solar System Barycenter");
const SpaceObject SpaceObject::Sun = SpaceObject(SUN_SPICE_ID, "Sun");
